### ISSUE TITLE
Rework Filtering Functions

### ISSUE THREAD

**Issue by szczura** [RickLeinecker/SeeShells#120]
See the in person meetings pictures for details.

**Comment by klaki892** [comment]
C&P'ed from team discussion for referencing later
-------------------------------------------------------
So as i was making the changes to the filters last night it dawned on me:
- Why are we filtering IEvents and not the Timeline graphical objects?
So here's my proposal: 
Instead of the `IFilter.apply()` taking a `ref List<IEvent>` as a parameter, we switch this for the `List<PrettyObject>` that exists inside `PrettyObjectCollection`
then, `PrettyObjectCollection` Implements the two functions `addFilter()` and `removeFilter()` that are currently about to go into `EventCollection`.
What the  functions do: is 
```for every `PrettyObject`
 look at it's `IEvent` inside of it, 
if the filter says it does **_NOT_** match,
 go to the `PrettyObject.Dot` and set its visibility to `false`. (or similar nature)
```
if a filter is cleared from the filterlist, restore the visibility property of all `PrettyObjects` in the list.
Why this would improve the existing idea:
1. Right now everytime we apply a filter, its up to us to not only generate a new copy of the event List ( `filteredEventList`) but to also tell the `PrettyObjectCollection` that you have a brand new list to completely generate new `PrettyObjects` for. Making new lists of references isnt expensive (`eventList` -> `filteredEventList`), but generating new "Pretty Objects", which have the `Block` and `Dot` is  more expensive since its also GUI rendering on top of the IEvents. ( `filteredEventlist` -> `prettyObjectList`)
2. right now we have a disconnect between updating `EventCollection` and `PrettyObjectCollection`  as it currently stands, for `PrettyObjectCollection` to get the items to render, `EventCollection` has to call a specific `PrettyObjectCollection` method each time the list updates. And this happens *after* a UI element tells `EventColletion` to add a new filter. By using this proposed method, the UI interacts with the UI (`UI Element.onChange` -> `PrettyObjectCollection`)   and keeps everything grouped to frontend specific elements. encapsulating changes to only PrettyObjectCollection and not its dependency (EventCollection) 
Tradeoffs/complexity changes:
 [Given N = number of events in the original `eventList` inside `EventCollection`] We go from the sometimes repeated O(N) operation of generating a `filteredEventList` each time a filter is removed, and a O(N-x) when adding a filter, to a guaranteed O(N) operation of  pretty objects iteration each time a filter is added or removed.  However, we change the current generation of  the PrettyObject list time from O(N) everytime we get a new filter list(O(n-x)) to O(1) because we never regenerate the list of `prettyObjects`, only act on the once-made-done list of `PrettyObjects` 
(F= number of times the filterlist changes)
AKA 
current: [eventlist generation O(N) + prettyObjectList generation O(N) + [filteredEventList generation O(N-x) + prettyObjectListGeneration O(N-x) ...  = O(F * 2(N-x)) ] == **O(2N) + O(F * 2(N-x))** 
Proposed: [eventList generation O(N) + prettyObjectList generation O(N) + [prettyObjectList filtered O(N) + ... ]== __**O(2N) + O(F * (N-x))**__
----
**TL;DR: If we filter by using the visibility property of the GUI events, we do less work overall and we change a 3 party communication system (UI -> EventCollection -> PrettyObjectCollection) to a 2 party system (UI -> PrettyObjectCollection)**
